### ISSUE TITLE
Efficiently support `for c in u"asdasd"[1:5]`

### ISSUE THREAD

**Issue by gsnedders** [cython/cython#3536]
Currently this generates a call to `__Pyx_PyUnicode_Substring` followed by `__Pyx_init_unicode_iteration`; notably the former allocates a new string for the substring, which in this case we have no actual need of: it should be possible to run this without allocating a new string.

**Comment by scoder** [comment]
Right. We should handle `SliceIndexNode` and `SliceNode` in `_transform_unicode_iteration()` in `Optimze.py`, similar to `_transform_carray_iteration()`. PR welcome.

**Comment by gsnedders** [comment]
I can't actually trigger the `_transform_carray_iteration` case in 0.15 at least, hmm. Always getting calls to `__Pyx_PyBytes_FromStringAndSize`. Do you have any example of where it triggers?

e.g.:

```
cdef char foo(char* s, char cmp_c, Py_ssize_t start, Py_ssize_t stop):
    cdef char c
    for c in s[start:stop]:
        if c == cmp_c:
            return c
    return 0
```

Copying the optimization is something I could do quickly, trying to figure out why that isn't working is probably orders of magnitude harder. :)

**Comment by scoder** [comment]
There are a couple of test cases in `for_in_string.pyx` and `carray_slicing.pyx`.

**Comment by gsnedders** [comment]
OK, so the current optimization seems to only be hit by (e.g.) `int*` and not `char*` (because the `char*` ends up typed as `bytes`).

**Comment by gsnedders** [comment]
I seem to now be working on this, starting off with the `char*` case (which I… very unsafely have working).
